### DISCUSSION TITLE
DNS resolver setup difficult, what about reserving 1 carrier-grade-NAT-shared IP for self

### DISCUSSION THREAD

**Discussion by gene1wood** [netbirdio/netbird#634] [Ideas & Feature Requests]
The new private DNS feature is great. The challenge is that Netbird clients must each configure their network stack to use the private DNS resolver, and for each client, this resolver is listening on a different IP (the IP of the Netbird interface).

As I see it, the result is that the only way to configure a Netbird client's network stack to use the private DNS resolver is to first discover the local Netbird IP address (under Linux, the IP of the `wt0` interface).

It would be great if this was easier and didn't require this discovery step.

One idea (which I don't know if it makes good sense), would be if Netbird allocated one of the [IPv4 Shared Address Space](https://en.wikipedia.org/wiki/IPv4_shared_address_space) IP addresses as a similar idea of the localhost loopback network ([127.0.0.0/8](https://en.wikipedia.org/wiki/Localhost)). The idea being, that on every Netbird client, if you initiated a connection to this reserved IP address (e.g. `100.127.255.254`) it would route to whatever the local Netbird client was.

If this new feature was possible, then configuring private DNS would be simple as every client would just have, as an example, `100.127.255.254` set as the DNS resolver.

I imagine having a fixed localhost-like address for Netbird would have other benefits for users as well.

If there's a simpler or other solution to the specific "how to setup a DNS resolver" question, do please share.

**Comment by gene1wood** [comment]
@TechHutTV I'm assuming there isn't an interest at Netbird in exploring this (based on it being closed). Did anyone actually look at this?

**Reply by jnfrati** [reply]
Hey @gene1wood! Apologies for that, just re-opened this feature request, we just assumed all this was staled as it was from 2022 :bow: 

Going to take a deeper look at this tomorrow :+1:

**Reply by gene1wood** [reply]
Ahh thanks! It's all good if this suggestion doesn't make sense, feel free to close, I just was hoping somebody would look at it.

**Reply by haydenstith** [reply]
Forgive me if I am misunderstanding the feature request but I believe this is already the current behavior for DNS resolution.
https://docs.netbird.io/manage/dns#dns-setup-per-platform
<img width="886" height="586" alt="image" src="https://github.com/user-attachments/assets/1dc58bcf-a355-4a25-9bf9-d31725367d02" />

**Reply by gene1wood** [reply]
When I suggested this back in December 2022, Netbird's behavior was to have the DNS listener listen on the Netbird client IP, or 127.0.0.1 or 127.0.0.153.

Later, in #902 this was changed to the current behavior that your screenshot shows (which was documented in https://github.com/netbirdio/docs/pull/504 ).

The current behavior still requires a discovery step (of determining the value of `x`) so it can't be configured in `/etc/resolv.conf` (because the IP is variable, depending on the second octet of the Netbird /16 block)

The suggestion for having a static/predictable IP address for the DNS resolver is meant to enable configuring clients to use that DNS resolver without having to dynamically figure out what the resolver IP is each time.

But yes, feel free to close if this isn't desired or interesting.
